File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        int stack_top = 10;
        char stack[10000];
        int sum = 0;
        for (int i = 0; i < s.size(); i++)
        {
            switch (s[i])
            {
            case '(':
            case '[':
            case '{':
                stack[stack_top] = s[i];
                stack_top++;
                sum++;
                break;
            case ')':
                if (stack[stack_top - 1] == '(')
                {
                    stack[stack_top] = '\0';
                    stack_top--;
                    sum--;
                }
                else
                {
                    return false;
                }
                break;
            case ']':
                if (stack[stack_top - 1] == '[')
                {
                    stack[stack_top] = '\0';
                    stack_top--;
                    sum--;
                }
                else
                {
                    return false;
                }
                break;
            case '}':
                if (stack[stack_top - 1] == '{')
                {
                    stack[stack_top] = '\0';
                    stack_top--;
                    sum--;
                }
                else
                {
                    return false;
                }
                break;
            }
        }
        if (sum != 0)
            return false;
        return true;
    }
};

File: Python/0020-valid-parentheses.py
# time complexity: O(n)
# space complexity: O(n)
class Solution:
    def isValid(self, s: str) -> bool:
        bracketMap = {"(": ")", "[": "]", "{": "}"}
        openSet = set(["(", "[", "{"])
        stack = []
        for char in s:
            if char in openSet:
                stack.append(char)
            elif stack and char == bracketMap[stack[-1]]:
                stack.pop()
            else:
                return False
        return stack == []


Input = "(())"
print(Solution().isValid(Input))
